beamer/chain.py

```diff
@@ -150,7 +150,7 @@ def add_events(self, events: list[Event]) -> None:
     def _thread_func(self) -> None:
         self._log.info("EventProcessor started")
         while not self._stop:
-            if self._have_new_events.wait(1):
+            if self._have_new_events.wait(1) or len(self._events) > 0:
```

Comment:
Is this correct? Couldn't it be the case that there are still events in the queue which cannot be processed? Wouldn't we in that case never wait anymore?

Reply:
In my understanding this will first wait, and then, if there are still events in the queue, process them, even if no new events came in.

Reply:
This would look like this:
```diff
diff --git a/beamer/chain.py b/beamer/chain.py
index 46fc26d..cd9d43a 100644
--- a/beamer/chain.py
+++ b/beamer/chain.py
@@ -150,7 +150,7 @@ class EventProcessor:
     def _thread_func(self) -> None:
         self._log.info("EventProcessor started")
         while not self._stop:
-            if self._have_new_events.wait(1) or len(self._events) > 0:
+            if self._have_new_events.wait(1):
                 self._have_new_events.clear()
                 self._process_events()

@@ -195,6 +195,8 @@ class EventProcessor:
             )
             iteration += 1
             if not any_state_changed:
+                if len(unprocessed) > 0:
+                    self._have_new_events.set()
                 break

         # New events might be created by event handlers. If they would be

Reply:
@fredo What's your favorite?